operational/vmware/instance_tag_sync/instance_tag_sync.pt

```diff
@@ -0,0 +1,452 @@
+  call sys_log("session_token:",$session_token)
+  # $category_id = "urn:vmomi:InventoryServiceCategory:f9c8236f-7320-47f0-8daa-1d7c83a673ab:GLOBAL"
+  call list_categories($param_wstunnel_token, $session_token) retrieve $category_ids
+  $category_id = null
+  foreach $category in $category_ids do
+    call get_category($param_wstunnel_token, $session_token, $category) retrieve $spec
+    if $spec["name"] == $param_tag_category
+      $category_id = $category
+    end
+  end
+
+  if $category_id == null
+    call create_category($param_wstunnel_token, $session_token,"MULTIPLE", "rightscale tag category", $param_tag_category) retrieve $category_id
+  end
+
+  $$tag_array = []
+  foreach $instance in $data do
+    call sys_log('VMWare Instance sync: ',to_s($instance))
+    $data_tags = $instance["tags"]
+    $$c_tag_ids = []
+    call list_tags_by_category($param_wstunnel_token, $session_token, $category_id) retrieve $tag_ids
+    $$c_tag_ids = $tag_ids
+    $$detach_response = []
+    if size($tag_ids) > 0
+      foreach $tag_id in $tag_ids do
+        call detach_tag($param_wstunnel_token, $session_token, $tag_id, $instance["id"]) retrieve $response
+        $$detach_response << $response
+      end
+    end
+    call list_tags($param_wstunnel_token, $session_token) retrieve $tag_ids
+    call sys_log("tag_ids:", to_s($tag_ids))
+    task_label('populating tags list')
+    foreach $tag_id in $tag_ids do
+      $match_tag = first(select($$tag_array, { "id": $tag_id }))
+      if $match_tag == null
+        call get_tag($param_wstunnel_token, $session_token, $tag_id) retrieve $spec
+        $$tag_array << { "name": $spec["name"], "id": $spec["id"] }
+      end
+    end
+    
+    foreach $tag in $data_tags do
+      task_label('instance:' + $instance["name"] + ', tag: '+ $tag)
+      $$matching_tag = []
+      $match_tag = first(select($$tag_array, { "name": $tag }))
+      if $match_tag == null
+        $$matching_id = null
+      else
+        call sys_log("matched tag:",$tag)
+        $$matching_tag << $match_tag
+        $$matching_id = $match_tag["id"]
+      end
+      if $$matching_id == null
+        call create_tag($param_wstunnel_token, $session_token, $category_id, "rightscale synced tag", $tag) retrieve $tag_id
+      else
+        $tag_id = $$matching_id
+      end
+      call attach_tag($param_wstunnel_token, $session_token, $tag_id, $instance["id"]) retrieve $response
+    end
+  end
+end
+
+# https://vdc-repo.vmware.com/vmwb-repository/dcr-public/1cd28284-3b72-4885-9e31-d1c6d9e26686/71ef7304-a6c9-43b3-a3cd-868b2c236c81/doc/operations/com/vmware/cis/session.create-operation.html
+define generate_session_token($tunnel_token, $username, $password) return $token do
+  $$str = $username + ":" + $password
+  $$encoded_auth = to_base64($$str)
+  $response = http_post(
+    insecure: true,
+    url: "https://wstunnel10-1.rightscale.com/_token/"+$tunnel_token+"/rest/com/vmware/cis/session",
```

Comment:
Is there only `wstunnel10-1` or should this be selectable somehow?

Reply:
I updated it to be an input, as well as set a default value, as I think most clients use 10-1 since that is what RCA-V defaults to.